src/provider/provider_os_memory.c

```diff
@@ -509,15 +599,179 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
     return 0;
 }
 
-static int get_membind(os_memory_provider_t *provider, size_t size) {
+/// membbind_t - some
+typedef struct membind_t {
+    /// Bitmap representing the set of nodes to which memory will be bound
+    hwloc_bitmap_t bitmap;
+    /// Size of the memory bound to the current node
+    size_t bind_size;
+    /// Address of the memory for next bind
+    char *addr;
+    /// Total size of memory allocation left to bind
+    size_t alloc_size;
+
+    /// Current node index
+    unsigned node;
+
+    /// Size of a single memory page
+    size_t page_size;
+
+    /// Remainder from the division used to distribute pages across nodes
+    size_t rest;
+
+    /// Number of pages to allocate
+    size_t pages;
+
+    /// Pages left to bind in current node
+    size_t leftover_bind;
+} membind_t;
+
+/// Advances the memory binding configuration for the next set of pages
+/// If we have to bind bytes which belongs to single page to mutiliple nodes,
+/// we will bind it to all nodes that those bytes belongs to - and lets kernel decide where to allocate it.
+static void nextBind(os_memory_provider_t *provider, membind_t *membind) {
+
+    // If all nodes have been processed.
+    if (membind->node == provider->partitions_len) {
+        // if alloc_size is not 0, it means that something is wrong
+        assert(membind->alloc_size == 0);
+        return;
+    }
+
+    // Reset the bitmap for next binding
+    hwloc_bitmap_zero(membind->bitmap);
+
+    // Flag to check if binding crosses partition boundaries
+    int bind_border_page = 0;
+    if (membind->leftover_bind != 0) {
+        // if we have move than page leftover from previous bind
```

Comment:
`than page` -> `than a page` (pls apply 3x in this file)

Reply:
memory binding